validator/src/journal/chain.rs

```diff
@@ -0,0 +1,683 @@
+
+use protobuf;
+
+use batch::Batch;
+use journal;
+use journal::block_wrapper::BlockWrapper;
+use metrics;
+
+use proto::transaction_receipt::TransactionReceipt;
+use scheduler::TxnExecutionResult;
+
+const RECV_TIMEOUT_MILLIS: u64 = 100;
+
+lazy_static! {
+    static ref COLLECTOR: metrics::MetricsCollectorHandle =
+        metrics::get_collector("sawtooth_validator.chain");
+}
+
+#[derive(Debug)]
+pub enum ChainControllerError {
+    QueueRecvError(RecvError),
+    ChainIdError(io::Error),
+    ChainUpdateError(String),
+    BlockValidationError(ValidationError),
+    BrokenQueue,
+}
+
+impl From<RecvError> for ChainControllerError {
+    fn from(err: RecvError) -> Self {
+        ChainControllerError::QueueRecvError(err)
+    }
+}
+
+impl From<io::Error> for ChainControllerError {
+    fn from(err: io::Error) -> Self {
+        ChainControllerError::ChainIdError(err)
+    }
+}
+
+impl From<ValidationError> for ChainControllerError {
+    fn from(err: ValidationError) -> Self {
+        ChainControllerError::BlockValidationError(err)
+    }
+}
+
+pub trait ChainObserver: Send + Sync {
+    fn chain_update(&mut self, block: &BlockWrapper, receipts: &[&TransactionReceipt]);
+}
+
+pub trait ExternalLock: Send + Sync {
+    fn lock(&self) -> Box<ExternalLockGuard>;
+}
+
+pub trait ExternalLockGuard: Drop {
+    fn release(&self);
+}
+
+pub trait ChainHeadUpdateObserver: Send + Sync {
+    /// Called when the chain head has updated.
+    ///
+    /// Args:
+    ///     block: the new chain head
+    ///     committed_batches: all of the batches that have been committed
+    ///         on the given fork. This may be across multiple blocks.
+    ///     uncommitted_batches: all of the batches that have been uncommitted
+    ///         from the previous fork, if one was dropped.
+    fn on_chain_head_updated(
+        &mut self,
+        block: BlockWrapper,
+        committed_batches: Vec<Batch>,
+        uncommitted_batches: Vec<Batch>,
+    );
+}
+
+pub trait BlockCache: Send + Sync {
+    fn contains(&self, block_id: &str) -> bool;
+
+    fn put(&mut self, block: BlockWrapper);
+
+    fn get(&self, block_id: &str) -> Option<BlockWrapper>;
+}
+
+#[derive(Debug)]
+pub enum ChainReadError {
+    GeneralReadError(String),
+}
+
+pub trait ChainReader: Send + Sync {
+    fn chain_head(&self) -> Result<Option<BlockWrapper>, ChainReadError>;
+    fn count_committed_transactions(&self) -> Result<usize, ChainReadError>;
+}
+
+#[derive(Debug)]
+pub enum ValidationError {
+    BlockValidationFailure(String),
+}
+
+pub trait BlockValidator: Send + Sync {
+    fn in_process(&self, block_id: &str) -> bool;
+    fn in_pending(&self, block_id: &str) -> bool;
+    fn validate_block(&self, block: BlockWrapper) -> Result<(), ValidationError>;
+
+    fn submit_blocks_for_verification(
+        &self,
+        blocks: &[BlockWrapper],
+        response_sender: Sender<(bool, BlockValidationResult)>,
+    );
+}
+
+// This should be in a validation module
```

Comment:
Any reason not to just do this now?

Reply:
The comment is from before rebasing on top of other changes, so I will look and see if it is possible now